File: src/cmidas/parser/lexer.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#include "lexer.h"

#define T Lexer_T

#define is_at_end(LEX) ((*(LEX)->index) == '\0')

#define is_alpha(C) (((C) >= 'a' && (C) <= 'z') || \
                     ((C) >= 'A' && (C) <= 'Z') || \
                      (C) == '_')

#define is_numeric(C) ((C) >= '0' && (C) <= '9')

#define is_alpha_num(C) (is_alpha(C) || is_numeric(C))

struct T
{
    const char *buffer;
    const char *index;
    const char *start;
           int  lineno;
           int  colno;
          bool  had_error;
          char  error_msg[256];
};

struct keyword
{
    const char *name;
    int length;
    enum tok_t type;
} keywords[] =
{
    { "and"  , 4, TOK_AND   },
    { "break", 6, TOK_BREAK },
    { "do"   , 3, TOK_DO    },
    { "else" , 5, TOK_ELSE  },
    { "end"  , 4, TOK_END   },
    { "false", 6, TOK_FALSE },
    { "for"  , 4, TOK_FOR   },
    { "fun"  , 4, TOK_FUN   },
    { "if"   , 3, TOK_IF    },
    { "or"   , 3, TOK_OR    },
    { "print", 6, TOK_PRINT },
    { "true" , 5, TOK_TRUE  },
    { "var"  , 4, TOK_VAR   },
    { "while", 6, TOK_WHILE },
    { NULL   , 0, TOK_EOF   }
};

static struct tok *tok_new   (T lex, enum tok_t type);
static struct tok *identifier(T lex);
static struct tok *number    (T lex);
static struct tok *string    (T lex);

static char char_next   (T lex);
static char lookahead   (T lex);
static bool char_matches(T lex, const char c);

static void skip_space(T lex);
static void skip_line (T lex);

T
Lex_new(void)
{
    T lex;

    lex = malloc(sizeof(struct T));
    lex->lineno = 1;
    lex->colno  = 0;

    return lex;
}

void
Lex_feed(T lex, const char *buffer)
{
    lex->buffer = buffer;
    lex->index  = buffer;
    lex->start  = buffer;
}

struct tok *
Lex_get_tok(T lex)
{
    char c;

    skip_space(lex);
    lex->start = lex->index;
    c = char_next(lex);
    if (is_alpha(c))
        return identifier(lex);

    if (is_numeric(c))
        return number(lex);

    switch (c)
    {
        case '!':
            if (char_matches(lex, '='))
                return tok_new(lex, TOK_BANG_EQUAL);
            else
                return tok_new(lex, TOK_BANG);

        case ',':
            return tok_new(lex, TOK_COMMA);

        case '=':
            if (char_matches(lex, '='))
                return tok_new(lex, TOK_EQUAL_EQUAL);
            else
                return tok_new(lex, TOK_EQUAL);

        case '>':
            if (char_matches(lex, '='))
                return tok_new(lex, TOK_GREAT_EQUAL);
            else
                return tok_new(lex, TOK_GREAT);

        case '#':
            skip_line(lex);
            /* XXX Eliminate recursion from lexer */
            return Lex_get_tok(lex);

        case '<':
            if (char_matches(lex, '='))
                return tok_new(lex, TOK_LESS_EQUAL);
            else
                return tok_new(lex, TOK_LESS);

        case '-':
            return tok_new(lex, TOK_MINUS);

        case '(':
            return tok_new(lex, TOK_PAREN_LEFT);

        case ')':
            return tok_new(lex, TOK_PAREN_RIGHT);

        case '%':
            return tok_new(lex, TOK_PERCENT);

        case '+':
            if (char_matches(lex, '+'))
                return tok_new(lex, TOK_PLUS_PLUS);
            else
                return tok_new(lex, TOK_PLUS);

        case '"':
            return string(lex);

        case ';':
            return tok_new(lex, TOK_SEMICOLON);

        case '/':
            if (char_matches(lex, '/'))
                return tok_new(lex, TOK_SLASH_SLASH);
            else
                return tok_new(lex, TOK_SLASH);

        case '*':
            return tok_new(lex, TOK_STAR);

        case '\0':
            return tok_new(lex, TOK_EOF);

        default:
            sprintf
            (
                lex->error_msg,
                "Line: %i\n"
                "Col : %i\n"
                "Encoutered unknown glyph `%c`.",
                lex->lineno, lex->colno, c
            );
            lex->had_error = true;
            return NULL;
    }
}

void
Lex_get_err(T lex)
{
    printf("%s\n", lex->error_msg);
}

void
print_tok(struct tok *tok)
{
    printf("<\"%s\", type %i, len %i, line %i, col %i>\n",
        tok->lexeme, tok->type, tok->length, tok->lineno, tok->colno);
}

static struct tok *
tok_new(T lex, enum tok_t type)
{
    struct tok *tok;
    char *lexeme;
    int length;

    length = lex->index - lex->start;

    tok = malloc(sizeof(struct tok));
    lexeme = malloc(length + 1);
    strncpy(lexeme, lex->start, length);
    lexeme[length] = '\0';

    tok->lexeme = lexeme;
    tok->length = length + 1;
    tok->type   = type;
    tok->lineno = lex->lineno;
    tok->colno  = lex->colno - length;

    return tok;
}

static struct tok *
identifier(T lex)
{
    struct tok *tok;
    struct keyword *keyw;

    while (is_alpha_num(lookahead(lex)))
        char_next(lex);

    tok = tok_new(lex, TOK_IDENTIFIER);
    for (keyw = keywords; keyw->name != NULL; ++keyw)
    {
        if (tok->length == keyw->length &&
            strcmp(tok->lexeme, keyw->name) == 0)
        {
            tok->type = keyw->type;
            break;
        }
    }

    return tok;
}

static struct tok *
number(T lex)
{
    bool is_double = false;

    for (;;)
    {
        if (is_numeric(lookahead(lex)))
            char_next(lex);
        else if (lookahead(lex) == '.')
        {
            if (is_double == false)
            {
                is_double = true;
                char_next(lex);
            }
            else
                break;
        }
        else
            break;
    }

    if (is_double)
        return tok_new(lex, TOK_DOUBLE);
    else
        return tok_new(lex, TOK_INTEGER);
}

static struct tok *
string(T lex)
{
    struct tok *tok;

    lex->start = lex->index;
    while (lookahead(lex) != '"')
        char_next(lex);

    tok = tok_new(lex, TOK_STRING);
    char_next(lex);

    return tok;
}

static char
char_next(T lex)
{
    char next;

    next = *lex->index;
    ++lex->index;
    ++lex->colno;

    return next;
}

static char
lookahead(T lex)
{
    return *lex->index;
}

static bool
char_matches(T lex, const char c)
{
    if (is_at_end(lex))
        return false;
    else if (*lex->index == c)
    {
        char_next(lex);
        return true;
    }
    else
        return false;
}

static void
skip_space(T lex)
{
    char c;
    for (;;)
    {
        c = lookahead(lex);
        switch (c)
        {
            case ' ':
            case '\r':
            case '\t':
                char_next(lex);
                break;

            case '\n':
                ++lex->lineno;
                lex->colno = 0;
                char_next(lex);
                break;

            default:
                return;
        }
    }
}

static void
skip_line(T lex)
{
    char c;
    for (;;)
    {
        c = lookahead(lex);
        switch (c)
        {
            case '\n':
                ++lex->lineno;
                lex->colno = 0;
                char_next(lex);
                return;

            case '\0':
                return;

            default:
                char_next(lex);
                break;
        }
    }
}


#undef T
#undef is_at_end
#undef is_alpha
#undef is_numeric
#undef is_alpha_num

File: src/pymidas/lexer.py
from toks import *

keywords = {
    "and"   : TOK_AND,
    "break" : TOK_BREAK,
    "do"    : TOK_DO,
    "else"  : TOK_ELSE,
    "end"   : TOK_END,
    "false" : TOK_FALSE,
    "for"   : TOK_FOR,
    "fun"   : TOK_FUN,
    "if"    : TOK_IF,
    "nil"   : TOK_NIL,
    "or"    : TOK_OR,
    "print" : TOK_PRINT,
    "return": TOK_RETURN,
    "true"  : TOK_TRUE,
    "var"   : TOK_VAR,
    "while" : TOK_WHILE
}


def is_alpha(char: str) -> bool:
    return ("a" <= char <= "z") or \
           ("A" <= char <= "Z") or\
            char == "_"


def is_numeric(char: str) -> bool:
    return "0" <= char <= "9"


def is_alpha_num(char: str) -> bool:
    return is_alpha(char) or is_numeric(char)


class Token():
    def __init__(self, lexeme, kind, length, lineno, colno):
        self.lexeme = lexeme
        self.kind = kind
        self.length = length
        self.lineno = lineno
        self.colno = colno

    def __repr__(self, bare=True):
        if bare:
            return self.lexeme
        else:
            return "(%s @ %i %i)" % (self.lexeme, self.lineno, self.colno)


class LexerError(Exception):
    pass


class Lexer():
    def __init__(self):
        self.buffer = None
        self.index = 0
        self.start = 0
        self.lineno = 1
        self.colno = 0

    def feed(self, buf):
        self.buffer = buf

    def char_next(self) -> str:
        char = self.buffer[self.index]
        self.index += 1
        self.colno += 1
        return char

    def lookahead(self) -> str:
        return self.buffer[self.index]

    def is_at_end(self) -> bool:
        return self.buffer[self.index] == '\0'

    def char_matches(self, char: str) -> bool:
        if self.is_at_end():
            return False
        elif self.buffer[self.index] == char:
            self.char_next()
            return True
        else:
            return False

    def skip_space(self):
        while True:
            char = self.lookahead()
            if char in " \r\t":
                self.char_next()
            elif char == "\n":
                self.lineno += 1
                self.colno = 0
                self.char_next()
            else:
                break

    def skip_line(self):
        while True:
            char = self.lookahead()
            if char == "\n":
                self.lineno += 1
                self.colno = 0
                self.char_next()
                break
            elif char == '\0':
                break
            else:
                self.char_next()

    def get_tok(self) -> Token:
        self.skip_space()
        self.start = self.index
        char = self.char_next()

        if is_alpha(char):
            return self.identifier()
        if is_numeric(char):
            return self.number()

        if char == "!":
            if self.char_matches("="):
                return self.tok(TOK_BANG_EQUAL)
            else:
                return self.tok(TOK_BANG)
        elif char == ",":
            return self.tok(TOK_COMMA)
        elif char == "=":
            if self.char_matches("="):
                return self.tok(TOK_EQUAL_EQUAL)
            else:
                return self.tok(TOK_EQUAL)
        elif char == ">":
            if self.char_matches("="):
                return self.tok(TOK_GREAT_EQUAL)
            else:
                return self.tok(TOK_GREAT)
        elif char == "#":
            self.skip_line()
            return self.get_tok()
        elif char == "<":
            if self.char_matches("="):
                return self.tok(TOK_LESS_EQUAL)
            else:
                return self.tok(TOK_LESS)
        elif char == "-":
            return self.tok(TOK_MINUS)
        elif char == "(":
            return self.tok(TOK_PAREN_LEFT)
        elif char == ")":
            return self.tok(TOK_PAREN_RIGHT)
        elif char == "%":
            return self.tok(TOK_PERCENT)
        elif char == "+":
            if self.char_matches("+"):
                return self.tok(TOK_PLUS_PLUS)
            else:
                return self.tok(TOK_PLUS)
        elif char == "\"":
            return self.string()
        elif char == ";":
            return self.tok(TOK_SEMICOLON)
        elif char == "/":
            if self.char_matches("/"):
                return self.tok(TOK_SLASH_SLASH)
            else:
                return self.tok(TOK_SLASH)
        elif char == "*":
            return self.tok(TOK_STAR)
        elif char == "\0":
            return self.tok(TOK_EOF)
        else:
            raise LexerError("Unknown character: '%s'" % char)

    def tok(self, kind) -> Token:
        lexeme = self.buffer[self.start:self.index]
        length = self.index - self.start
        return Token(lexeme, kind, length, self.lineno, self.colno - length)

    def identifier(self) -> Token:
        while is_alpha_num(self.lookahead()):
            self.char_next()

        tok = self.tok(TOK_IDENTIFIER)
        if tok.lexeme in keywords:
            tok.kind = keywords[tok.lexeme]

        return tok

    def number(self) -> Token:
        is_double = False

        while True:
            if is_numeric(self.lookahead()):
                self.char_next()
            elif self.lookahead() == ".":
                if not is_double:
                    is_double = True
                    self.char_next()
                else:
                    break
            else:
                break

        if is_double:
            return self.tok(TOK_DOUBLE)
        else:
            return self.tok(TOK_INTEGER)

    def string(self) -> Token:
        self.start = self.index
        while self.lookahead() != "\"":
            self.char_next()
        tok = self.tok(TOK_STRING)
        self.char_next()
        return tok
